### PR TITLE
Send notification emails to job alert subscribers when alert is going to end

### PR DESCRIPTION
**Pull request by pezholio**
## Trello card URL:

https://trello.com/c/WVP6sEtD/849-send-notification-emails-to-job-alert-subscribers-when-alert-is-going-to-end-after-3-months

## Changes in this PR:

This adds two reminders to be sent one week and one day before a subscription is due to expire. Users click on a link and are then taken to a page where they can review their subscription and resubscribe.

## Screenshots of UI changes:

image

image

image

image

## Next steps:

- [x] Terraform deployment required?

- [ ] New development configuration to be shared?

### CODE REVIEW

**Comment by pezholio** [comment]
[Send notification emails to job alert subscribers when alert is going to end after 3 months](https://trello.com/c/WVP6sEtD/849-send-notification-emails-to-job-alert-subscribers-when-alert-is-going-to-end-after-3-months)

File: spec/jobs/send_first_subscription_expiry_alerts_job_spec.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    expect { job }.to change(ActiveJob::Base.queue_adapter.enqueued_jobs, :size).by(1)
+  end
+
+  it 'is in the send_expiry_alerts queue' do
+    expect(job.queue_name).to eq('send_expiry_alerts')
+  end
+
+  context 'when job is run' do
+    let!(:expiring_next_week) { create_list(:subscription, 3, expires_on: 1.week.from_now) }
+    let!(:expiring_next_month) { create_list(:subscription, 3, expires_on: 1.month.from_now) }
```
Comment (erbridge) [inline spec/jobs/send_first_subscription_expiry_alerts_job_spec.rb]: Let's make these explicitly `.to_date`?

File: spec/models/subscription_spec.rb [hunk 2]
```diff
@@ -224,4 +260,22 @@
       it { expect(subscription.expired?).to eq(true) }
     end
   end
+
+  describe 'expiring_in_7_days' do
+    let!(:expiring_next_week) { create_list(:subscription, 3, expires_on: 1.week.from_now) }
+    let!(:expiring_next_month) { create_list(:subscription, 3, expires_on: 1.month.from_now) }
```
Comment (erbridge) [inline spec/models/subscription_spec.rb]: Let's make these explicitly `.to_date`?

File: spec/models/subscription_spec.rb [hunk 3]
```diff
@@ -224,4 +260,22 @@
...
+
+    it 'gets all subscriptions expiring in 7 days' do
+      expect(Subscription.expiring_in_7_days).to eq(expiring_next_week)
+    end
+  end
+
+  describe 'expiring_tomorrow' do
+    let!(:expiring_next_week) { create_list(:subscription, 3, expires_on: 1.week.from_now) }
+    let!(:expiring_tomorrow) { create_list(:subscription, 3, expires_on: 1.day.from_now) }
```
Comment (erbridge) [inline spec/models/subscription_spec.rb]: Let's make these explicitly `.to_date`?

File: app/controllers/subscriptions_controller.rb [hunk 4]
```diff
@@ -25,23 +28,39 @@ def create
     render 'new'
   end
 
+  def renew
+    subscription = Subscription.find_or_initialize_by_token(params[:subscription_id])
+    @subscription = SubscriptionPresenter.new(subscription)
```
[the remark this answers is not in the document]
Comment (Renny4Real) [reply app/controllers/subscriptions_controller.rb]: @erbridge 
`Could this also be using fetch_subscription_from_token?`
This doesn’t seem to be the case.
We’ve had an attempt at using the `fetch_subscription_from_token` method. however the subscription can’t be found when by using `find_and_verify_by_token` (which is being implemented within `fetch_subscription_from_token`).
It seems like the renew action works for two use cases: Renewing a subscription prior to expiry (e.g user clicks link before expiry) and renewing after a subscription has expired (e.g user clicks link after expiry date).
I imagine that’s where the need for the `find_or_initialize_by_token` method came from.
We’d attempted the refactoring and the tests error on not being able to find a subscription (i.e `@subscription` is nil) which is why we feel it might a case of renew attending to two different use cases. Any suggestions or feedback on this though would be appreciated.

`Or could fetch_subscription_from_token add the presenter wrapper?`
Not sure how this would work.
The `unsubscribe` method doesn’t implement the presenter and `fetch_subscription_from_token` is a before action for this method. Would adding the presenter mean that the function of this method would have to change? Would also appreciate any suggestions or feedback on this though.

**Comment by erbridge** [comment]
Can we remove 5ee65ce with a rebase, please.

File: app/controllers/subscriptions_controller.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (erbridge) [reply app/controllers/subscriptions_controller.rb]: :+1:

File: terraform/modules/ecs/ecs.tf [hunk 5]
```diff
@@ -767,6 +798,23 @@ resource "aws_cloudwatch_event_rule" "send_job_alerts_daily_email_task" {
...
+  memory                   = "512"
+  execution_role_arn       = "${aws_iam_role.ecs_execution_role.arn}"
+  task_role_arn            = "${aws_iam_role.ecs_execution_role.arn}"
+}
+
+resource "aws_cloudwatch_event_rule" "send_expiry_alerts_task" {
+  name                = "${var.ecs_service_web_task_name}_send_expiry_alerts_task"
+  description         = "Send alerts to users who's alerts are due to expire"
+  schedule_expression = "${var.send_expiry_alerts_schedule}"
+}
```
Comment (erbridge) [inline terraform/modules/ecs/ecs.tf]: All this needs to be refactored using the new Terraform for declaring rake task definitions. Grab me if this isn't obvious.

File: terraform.tf [hunk 6]
```diff
@@ -94,6 +94,9 @@ module "ecs" {
   send_job_alerts_daily_email_command  = "${var.send_job_alerts_daily_email_command}"
   send_job_alerts_daily_email_schedule = "${var.send_job_alerts_daily_email_schedule}"
 
+  send_expiry_alerts_schedule = "${var.send_expiry_alerts_schedule}"
+  send_expiry_alerts_command = "${var.send_expiry_alerts_command}"
```
Comment (erbridge) [inline terraform.tf]: Nitpick: please run `terraform fmt` on this (`Format Document` does this in VSCode if you have the Terraform extension).

File: app/jobs/send_final_subscription_expiry_alerts_job.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+class SendFinalSubscriptionExpiryAlertsJob < ApplicationJob
+  queue_as :send_expiry_alerts
+
+  def perform
+    Subscription.due_final_expiry_notice.each do |s|
+      SubscriptionExpiryMailer.final_expiry_warning(s.id).deliver_later
+    end
```
Comment (erbridge) [inline app/jobs/send_final_subscription_expiry_alerts_job.rb]: Feels safer like this.

File: app/models/subscription.rb [hunk 8]
```diff
@@ -11,6 +11,10 @@ class Subscription < ApplicationRecord
   validates :search_criteria, uniqueness: { scope: %i[email expires_on frequency] }
 
   scope :ongoing, -> { where('expires_on >= current_date') }
+  scope :first_reminder_not_sent, -> { where(first_reminder_sent: false) }
+  scope :final_reminder_not_sent, -> { where(final_reminder_sent: false) }
+  scope :due_first_expiry_notice, -> { first_reminder_not_sent.where('expires_on <= ?', 1.week.from_now.to_date) }
+  scope :due_final_expiry_notice, -> { final_reminder_not_sent.where('expires_on <= ?', 1.day.from_now.to_date) }
```
Comment (erbridge) [inline app/models/subscription.rb]: Maybe we should exclude final expiries from first expiries? That is, users should only get one email regarding expiry on a given day. Though this should be an edge case, so happy to continue as is.

File: app/views/subscriptions/update.html.haml [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+.govuk-grid-row
+  .govuk-grid-column-two-thirds
+    .govuk-panel.govuk-panel--confirmation
+      %h1.govuk-panel__title= t('subscriptions.update.header')
+      .govuk-panel__body
+        = t('subscriptions.reference')
+        %br
+        %strong= @subscription.reference
+
+    %h3.govuk-heading-m=t('subscriptions.confirmation.next_step')
```
Comment (erbridge) [inline app/views/subscriptions/update.html.haml]: Can we duplicate the content into a new key `subscriptions.update.next_step`? I'm nervous about confirmation page content edits accidentally changing this content unintentionally.

File: app/views/subscriptions/update.html.haml [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    %p.govuk-body=t('subscriptions.update.next_step_details', email: @subscription.email, date: l(@subscription.expires_on))
+    %div.govuk-inset-text
+      %ul.govuk-list
+        - @subscription.filtered_search_criteria.each_pair do |filter, value|
+          %li
+            - if filter.present?
+              %span{ class: 'govuk-!-font-weight-bold'} #{filter.humanize}:
+            = value
```
Comment (erbridge) [inline app/views/subscriptions/update.html.haml]: This looks like the `criteria` partial, to me.

File: config/locales/en.yml [hunk 11]
```diff
@@ -275,11 +275,13 @@ en:
...
     confirmation:
-      header: Your email subscription has started
+      header: You have subscribed to a job alert
       receipt_confirmation: We have sent you a confirmation email.
       next_step: 'What happens next'
       unsubscribe: "You can unsubscribe from this subscription at any time by following the 'stop receiving emails for this search' link in the email."
-      next_step_details: 'We will run a search matching your search criteria and send a daily email with the results to %{email} for the next 6 months, until %{date}.'
+      next_step_details: 'You will be emailed details of any new jobs that match the following search criteria:'
+      continue: "You'll only be sent one job alert email per day, at the end of any day when 1 or more jobs match your criteria."
+      unsubscribe: This job alert subscription will end on %{date}. You can cancel it at any time using the link found at the bottom of job alert emails.
```
Comment (erbridge) [inline config/locales/en.yml]: This key is duplicated.

Comment (erbridge) [inline config/locales/en.yml]: I think the last sentence is a little unclear by what it means by `the bottom of job alert emails`. Suggest rewording to `the bottom of any job alert emails you receive`.

File: config/locales/en.yml [hunk 12]
```diff
@@ -275,11 +275,13 @@ en:
...
     reference: 'Your reference:'
     confirmation:
-      header: Your email subscription has started
+      header: You have subscribed to a job alert
       receipt_confirmation: We have sent you a confirmation email.
       next_step: 'What happens next'
       unsubscribe: "You can unsubscribe from this subscription at any time by following the 'stop receiving emails for this search' link in the email."
-      next_step_details: 'We will run a search matching your search criteria and send a daily email with the results to %{email} for the next 6 months, until %{date}.'
+      next_step_details: 'You will be emailed details of any new jobs that match the following search criteria:'
+      continue: "You'll only be sent one job alert email per day, at the end of any day when 1 or more jobs match your criteria."
```
Comment (erbridge) [inline config/locales/en.yml]: Can we be consistent with `1` vs `one`?

File: config/locales/en.yml [hunk 13]
```diff
@@ -296,6 +298,18 @@ en:
...
+      title: Resubscribe to job alert
+      intro: 'You are subscribed to email alerts for jobs matching the following criteria:'
+      expired: This job alert subscription has now ended
+      end_date: This job alert subscription ends on %{date}
+      continue: To continue receiving these alerts for another 6 months, just resubscribe at the bottom of this page.
+      change_instructions: "If you're resubscribing and want to change your email address or reference, just update them in the following fields:"
+    update:
+      header: You have resubscribed to a job alert
+      next_step_details: 'You will be emailed details of any new jobs that match the following search criteria:'
+      what_to_expect: You'll only be sent one job alert email per day, at the end of any day when 1 or more jobs match your criteria.
```
Comment (erbridge) [inline config/locales/en.yml]: Can we be consistent with `1` vs `one`?

File: config/locales/en.yml [hunk 14]
```diff
@@ -310,6 +324,16 @@ en:
             one: "A new job matching your search criteria was posted yesterday."
             other: "%{count} new jobs matching your search criteria were posted yesterday."
           unsubscribe: Follow this link to unsubscribe.
+    expiry:
+      email:
+        resubscribe: Resubscribe now
+        first_warning:
+          subject: "Job alert for '%{reference}' expiring - resubscribe now"
```
Comment (erbridge) [inline config/locales/en.yml]: *`is expiring`?

File: config/locales/en.yml [hunk 15]
```diff
@@ -310,6 +324,16 @@ en:
...
+          first_line: "Your job alert for your alert '%{reference}' on Teaching Vacancies will expire on %{expiry_date}. After this date you will no longer receive emails alerting you to new jobs that match these criteria."
```
Comment (erbridge) [inline config/locales/en.yml]: Extra `your alert` in here, I think. Should be `Your job alert for '%{reference}'`.

Comment (erbridge) [inline config/locales/en.yml]: We don't show the criteria, so referring to them seems confusing. Should we add the criteria to the warning email template?

File: config/locales/en.yml [hunk 16]
```diff
@@ -310,6 +324,16 @@ en:
...
+          reset_instructions: "You can resubscribe to this job alert for another 6 months by clicking the 'resubscribe' link below."
+        final_warning:
+          subject: "Job alert for '%{reference}' ends today - resubscribe now"
+          first_line: "Your job alert for your alert '%{reference}' on Teaching Vacancies has expired. If you would like to resubscribe to this alert for another 3 months click the 'resubscribe' link below. Otherwise you will no longer receive alerts for jobs matching these criteria."
```
Comment (erbridge) [inline config/locales/en.yml]: Again, we refer to the criteria without showing them to the user.

File: config/sidekiq.yml [hunk 17]
```diff
@@ -15,6 +15,7 @@
   - [audit_subscription_creation, 1]
   - [audit_spreadsheet, 1]
   - [seed_vacancies_from_api, 1]
+  - [send_expiry_alerts, 1]
```
Comment (erbridge) [inline config/sidekiq.yml]: Does this also need a concurrency setting?

File: spec/factories/subscriptions.rb [hunk 18]
```diff
@@ -4,6 +4,8 @@
     email { Faker::Internet.email }
     reference { Faker::Lorem.sentence }
     frequency { :daily }
+    first_reminder_sent { false }
+    final_reminder_sent { false }
```
Comment (erbridge) [inline spec/factories/subscriptions.rb]: Should this be unnecessary given the migration sets default values?

File: spec/mailers/subscription_spec.rb [hunk 19]
```diff
@@ -24,23 +24,26 @@
...
+    it 'sends a confirmation email' do
+      expect(mail.subject).to eq(I18n.t('job_alerts.confirmation.email.subject', reference: subscription.reference))
+      expect(mail.to).to eq([subscription.email])
+      expect(body_lines[0]).to match(/# #{I18n.t('app.title')}/)
+      expect(body_lines[1]).to match(/# #{subscription.reference}/)
+      expect(body_lines[5]).to match(/#{I18n.t('subscriptions.email.confirmation.subheading')}/)
+      expect(body_lines[7]).to match(/\* Subject: English/)
+      expect(body_lines[8]).to match(/\* Minimum Salary: £20,000/)
+      expect(body_lines[9]).to match(/\* Maximum Salary: £40,000/)
+      expect(body_lines[10]).to match(/\Suitable for NQTs/)
```
Comment (erbridge) [inline spec/mailers/subscription_spec.rb]: Is the `\` serving a purpose?

File: variables.tf [hunk 20]
```diff
@@ -163,6 +163,17 @@ variable "send_job_alerts_daily_email_schedule" {
   default     = "cron(0 08 * * ? *)"
 }
 
+variable "send_expiry_alerts_schedule" {
+    description = "send_expiry_alerts schedule expression - https://docs.aws.amazon.com/AmazonCloudWatch/latest/events/ScheduledEvents.html"
```
[the remark this answers is not in the document]
Comment (erbridge) [reply variables.tf]: I think this is still relevant isn't it?

File: app/views/subscriptions/update.html.haml (hunk 9 above)
Comment (pezholio) [reply app/views/subscriptions/update.html.haml]: Good shout 👍

**Comment by pezholio** [comment]
@erbridge All done now - I've ignored the copy ones for now as I imagine copy changes will come from Michael too, so don't want to change stuff that's likely going to get changed again

File: config/locales/en.yml (hunk 14 above)
Comment (erbridge) [reply config/locales/en.yml]: Or `expires soon` is better.

File: db/migrate/20190514080151_add_reminder_flags_to_subscription.rb [hunk 21]
```diff
@@ -0,0 +1,6 @@
+class AddReminderFlagsToSubscription < ActiveRecord::Migration[5.2]
```
Comment (erbridge) [inline db/migrate/20190514080151_add_reminder_flags_to_subscription.rb]: I think this will need its date made later because the working patterns changes had a more recent migration.

File: app/models/subscription.rb (hunk 8 above)
Comment (adrianclay) [reply app/models/subscription.rb]: TODO: Clarify if this is checking the penultimate day or the final day.

**Comment by adrianclay** [comment]
Closing for now, as deprioritised.
